### PR TITLE
plugin: Add dynamic support for CN version of FFXIV

### PR DESCRIPTION
**Pull request by KattTails**
Hiya,

As requested in #821 I've added support for the CN client, using memory offsets from the 0.11.2 release targetting FFXIV ver 5.01.

This also includes the changes from #824

### CODE REVIEW

File: CactbotOverlay/FFXIVProcessIntl.cs [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+
+      [FieldOffset(0x3E)]
+      public short level;
+
+      [FieldOffset(0x5F)]
+      public short shieldPercentage;
+    }
+    public FFXIVProcessIntl(ILogger logger) : base(logger) { }
+
+    [Serializable]
```
Comment (KattTails) [inline CactbotOverlay/FFXIVProcessIntl.cs]: FFXIVProcess.cs has dummy entries for all of the JobMemory structs and I've defined them in each version as although they're identical at the moment a future patch could change them and it saves work splitting them later on.

**Comment by ShadyWhite** [comment]
I am curious how it recognizes the CN client. If it is through the option of FFXIV_ACT_Plugin, this may not work, because FFXIV_ACT_Plugin does not officially support the CN client. We are using a modified version, which cannot identify the client as the "cn".

**Comment by KattTails** [comment]
The ACT plugin has an option for CN: 
image

**Comment by ShadyWhite** [comment]
Yes, the Chinese option was added in version 2.0.4.2, it works for 5.1 clients, not 5.01.  We are using a modified version 2.0.4.0.  But I don't think this is a problem, there is no need to modify your code, because I think this code can work with version 2.0.4.2 ACT Plugin after CN client update to 5.1.  On behalf of the Chinese players, I thank you very much for your efforts!

**Comment by quisquous** [comment]
Sorry, I merged the other one! Can you update this to not include it, and I can merge this one too?

I think the only other comment I have is that in the short term we could also check the ffxiv act dll version and if it's the old 2.0.4.0 then also use the CN version.

**Comment by KattTails** [comment]
> Sorry, I merged the other one! Can you update this to not include it, and I can merge this one too?

I had to wrangle git a bit, but I *think* I've managed it. 

> I think the only other comment I have is that in the short term we could also check the ffxiv act dll version and if it's the old 2.0.4.0 then also use the CN version.

I've added that check too, good idea!

**Comment by ShadyWhite** [comment]
I merged your codes from your repo to my fork and built a plugin.But still get the error of GetJobSpecificData.
image
image

**Comment by KattTails** [comment]
I just pushed a new build that ouputs which version it's targetting. @ShadyWhite, would you be kind enough to try this new build and check the output log to see which version it has detected? 

If it's still using the 'Intl' version, would you possibly be able to upload a copy of the modified FFXIV_ACT_Plugin so I can see what the version string is?
If it *is* using the Cn version, then the memory offsets are wrong somewhere, which I can try and fix instead.


Also as an aside, I just noticed I've implemented a different version checking method to the `VersionChecker` class, which I didn't see before now. Oops. I can remove that, @quisquous, although mine is a bit cleaner than using regexes!

**Comment by ShadyWhite** [comment]
The logs now shows like this.
image
The CN client mainly uses 3 versions of FFXIV_ACT_Plugin modified by 3 developers.
I have uploaded them to Google drive.https://drive.google.com/open?id=1Bzkso05VR9fJL_j2Y8-mafEr5A2Xc7LP

**Comment by KattTails** [comment]
Okay, well the version matches up with what I was testing for on all of those dlls so that's all fine, and in your log it is using the cn version. From your previous screenshot showing the cactbot test ui, the id/pos/rot fields seem to be working, meaning it's only likely this offset is wrong: 
```C#
[FieldOffset(0x18A4)]
public CharacterDetails charDetails;
```
I've gone ahead and written a quick build that will dump the info I need to work this out, which I'll upload here: [CactbotOverlay.zip](https://github.com/quisquous/cactbot/files/3923436/CactbotOverlay.zip). 

Every 10 seconds it will print out a long string of hex that looks like this: 
image

If you could kindly run this version, and then paste that entire line here, along with your class and hp / max hp, I'll be able to update this to work properly.

Let me know if you need anything and thank you for your help!

**Comment by ShadyWhite** [comment]
Strangely, your quick build works fine, it can read my job and HP correctly(ui/jobs works fine), and the data in test are all correct. I think the problem may be in my fork, sorry.
image
Then this is hex, but I think it may be useless for now.
Info: 78856fd4f77f00000000000000000000517203c30000a0411d4ef0400000000060a4cb3e0000000000000000000000005261697975690000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000899c1410000000000000000000000000000000e000002c000104000100010000bf41a0000000000000000000f7d908c3d277a8418537703f00000000806e513d0000803f0000a03fcdcc8c3f0000003f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000601308973f01000000000000000000000000000000000000000000000000000000000000c7c60f40d9cef7340000000000000000f2eaec3fd9cef7340000000000008037205caf3f00008036000000000000000000000000b0aa50a23f0100000000000000000000000000000000000000000000c7c60f40d9cef7340000000000000000f2eaec3fd9cef734000000000000000000000000000000000000000000886fd4f77f00000000000000000000e0e96dd4f77f000000000000000000005040729c3f0100000000000000000000d8e96dd4f77f00000000000000000000c0e96dd4f77f00000000000000000000000000000000000000000000000000000000000000000000ffffffff0000000000000000000000000000000000000000c8e96dd4f77f00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d0e96dd4f77f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000803f000000000000000000000000000000000000803f000000000000000000000000000000000000803f00000000000000000000000000000000000000000000803f000000000000000000000000000000000000803f000000000000000000000000000000000000803f0000000000000000000000000000000000000000b0e96dd4f77f00000000000000000000b0e96dd4f77f00000000000000000000b0e96dd4f77f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000f041729c3f0100000000000000000000000000000000000000000000000080bf000080bf000080bf000080bf000000000000000000000000000000000000000100000000000000000000000000000000d8e96dd4f77f00000000000000000000c0e96dd4f77f00000000000000000000000000000000000000000000000000000000000000000000ffffffff0000000000000000000000000000000000000000c8e96dd4f77f00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d0e96dd4f77f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000803f000000000000000000000000000000000000803f000000000000000000000000000000000000803f00000000000000000000000000000000000000000000803f000000000000000000000000000000000000803f000000000000000000000000000000000000803f0000000000000000000000000000000000000000b0e96dd4f77f00000000000000000000b0e96dd4f77f00000000000000000000b0e96dd4f77f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000f041729c3f0100000000000000000000000000000000000000000000000080bf000080bf000080bf000080bf010000000000000000000000000000000000000100000000000000000000000000000000d8e96dd4f77f00000000000000000000c0e96dd4f77f00000000000000000000000000000000000000000000000000000000000000000000ffffffff0000000000000000000000000000000000000000c8e96dd4f77f00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d0e96dd4f77f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000803f000000000000000000000000000000000000803f000000000000000000000000000000000000803f00000000000000000000000000000000000000000000803f000000000000000000000000000000000000803f000000000000000000000000000000000000803f0000000000000000000000000000000000000000b0e96dd4f77f00000000000000000000b0e96dd4f77f00000000000000000000b0e96dd4f77f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000f041729c3f0100000000000000000000000000000000000000000000000080bf000080bf000080bf000080bf02000000000000000000000000000000000000010000000000000000000000000000000036c26c3d030000000100000000000000105e6fd4f77f00000000000000000000385c6fd4f77f00000000000000000000305c6fd4f77f00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000305c6fd4f77f00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002048729c3f01000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000b05c6fd4f77f0000000000000000000000000000000000000000000000000000000000000000000000000000000000002048729c3f010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000080bf000080bf000080bf000080bf000080bf00000000000000000000000000000000000000000000000000000000185c6fd4f77f0000000000000000000000000000000000000000000000000000185c6fd4f77f000031a7b23eea6ab13e418a963e9a26613e05000000000000000000000000000000485c6fd4f77f000000000000000000002048729c3f010000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000285c6fd4f77f00002048729c3f010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000405c6fd4f77f0000204b729c3f010000105d6fd4f77f0000204b729c3f010000205c6fd4f77f000005200000052000000000000000000000f85d6fd4f77f0000583b34d23f01000000000000000000002048729c3f0100005040729c3f010000000000000000000000000000000080bf0000803f00000000896e2c3f000000000000000000000001000000000000000080e96dd4f77f0000000000000000000078e96dd4f77f0000000000000000000000000000000000000000000000000000000000000000803f000000000000000000000000000000000000000001000000ffffffffff0000000000000000000000706b42a23f0100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000700c0000000000000000000000000000000000000000000000000000700c00000000000000000000000000000000000000000000000000006f0c02000000000000000000000000005b0200000000000090007300700c00000000000000000000000000000000000000000000000000006f0c700c0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803f000000005040729c3f010000d0f8c4af3f010000020000000100000000000000000000000000000000000000a8e96dd4f77f0000000000000000000078e96dd4f77f0000000000000000000000000000000000000000000000000000000000000000803f0000000000000000000000000000000000000000ffffffffffffffffff0000000000000000000000a0e96dd4f77f0000000000000000000000000f031703140300000000000000000000000000001b0319031a03180300005040729c3f010000000000000000000000000000000000000000000000000000c8696fd4f77f00005040729c3f010000000000006c000000000000e0000000000004000000000000000000000000000090fd83d4f77f000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000078fd83d4f77f00000000000000000000000000000000000000007043000070436bc2533f00000000000000e000000000000000e0000000000000000000000000000000000000017f0a00000000000000000000e00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003000000f253729c3f010000400000000000000001000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003000000000000006a54729c3f01000040000000000000000100000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005040729c3f0100000000000000000000000000000000000000000000000000000000000000000000e9034c0001000000a08903973f010000e0d86dd4f77f0000a08903973f010000601308973f010000901893bb3f010000747332000900000005000000000000000100000001000000cdcc8c3fcdcc8c3f0000803f000000000000803f02000000030400000000000000000000000000000000000000000000e0d86dd4f77f000000000000000000000000000000000000000000000000000000000000ffffffff200000000000000000000000000000000000803f0000803f0000803f000000000000803f12000000010000000000000000000000000000000000000000000000e0d86dd4f77f000000000000000000000000000000000000000000000000000000000000ffffffff000000000000000000000000000000000000803f0000803f00000000000000000000803f12000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000803f0000803f15020100cb17016515020100150201001502010055000100350001006e0001003500010035000100000000000000000000000000000000000000000000000000000f0000000000000000000000000000000000e00000000000000000000000000000000000000000000000e0000000009000849c3f01000000000000000000009000849c3f0100000000000000000000505d12c93f0100000000000000000000070001640d020380993ba06c7707033b02010202026404000135e8b5a4e9acbc000000000000000000000000000000001bb20f00940000000000000000000000000000e000000000000000e00000000000000000000000002870463f000000000000000060c47cbf0000000000000000000000000000000000000000000000000000000000000000000000000000803f0000000000000000000000000000000000000000000000000000000000000000000000e0000000001304130400000000fd5300005c1701005c170100102700001027000000005906000000000000ca01000000008e00ffffffff030022000000000000000000000000000100190050001b000000010002000000000000000000030b00ffff000000000000000100000000000000
BLM 71516/71516

**Comment by KattTails** [comment]
Yes, it would seem the problem is in your fork if this build works fine. No worries, and thanks for all your help regardless!

**Comment by ShadyWhite** [comment]
Thanks a lot for your awesome work!

**Comment by quisquous** [comment]
Thanks for all this work.  :)

**Pull request merged**